lib/logstash/filters/kubernetes_metadata.rb

```diff
@@ -0,0 +1,290 @@
+      # in annotations
+      %w{ stderr stdout }.each do |t|
+        [ "log-format-#{t}-#{n}", "log-format-#{n}", "log-format-#{t}", "log-format" ].each do |k|
+          if v = a[k]
+            format[t] = v
+            break
+          end
+        end
+      end
+
+      metadata['log_format_stderr'] = format['stderr']
+      metadata['log_format_stdout'] = format['stdout']
+      @logger.debug("kubernetes metadata", metadata: metadata)
+
+    rescue => e
+      @logger.warn("Error setting log format. Please check log-format annotation.", error => e.to_s)
+    end
+  end
+
+  # based on https://github.com/vaijab/logstash-filter-kubernetes/blob/master/lib/logstash/filters/kubernetes.rb
+  def get_file_info(path)
+    parts = path.split(File::SEPARATOR).last.gsub(/.log$/, '').split('_')
+    if parts.length != 3 || parts[2].start_with?('POD-')
+      return nil
+    end
+    kubernetes = {}
+    kubernetes['replication_controller'] = parts[0].gsub(/-[0-9a-z]*$/, '')
+    kubernetes['pod'] = parts[0]
+    kubernetes['namespace'] = parts[1]
+    kubernetes['container_name'] = parts[2].gsub(/-[0-9a-z]*$/, '')
+    kubernetes['container_id'] = parts[2].split('-').last
+    return kubernetes
+  end
+
+  def sanitize_keys(data)
+    return {} unless data
+
+    parsed_data = {}
+    data.each do |k,v|
+      new_key = k.gsub(/\.|,/, '_')
+        .gsub(/\//, '-')
+      parsed_data[new_key] = v
+    end
+
+    return parsed_data
+  end
+
+  def get_kubernetes(namespace, pod)
+
+    begin
+      @logger.debug("Attempting to query the Kubernetes API")
+
+      url = [ @api, 'api/v1/namespaces', namespace, 'pods', pod ].join("/")
+
+      rest_opts = {
+        verify_ssl: @verify_api_ssl
+      }
+
+      if @auth_basic_user && @auth_basic_pass
+        @logger.debug("Found basic auth for Kubernetes API")
+
+        basic_user = @auth_basic_user
+        basic_pass = @auth_basic_pass
+
+        rest_opts.merge!( user: basic_user, password: basic_pass )
+      elsif @auth_bearer_key
+        @logger.debug("Found Bearer  auth for Kubernetes API")
+
+        bearer_key = @auth_bearer_key
+
+        rest_opts.merge!( Authorization: "Bearer #{bearer_key}" )
+      end
+
+      @logger.debug("rest_opts: #{rest_opts}")
+
+      begin
+        response = RestClient::Resource.new(url, rest_opts).get
+      rescue RestClient::ResourceNotFound
+        @logger.warn("Kubernetes returned an error while querying the API", url: url)
+      rescue Exception => e
+        @logger.warn("Error while querying the API", :error => e.to_s)
+      end
+
+      if response && response.code != 200
+        @logger.warn("Non 200 response code returned", code: response.code)
+      end
+
+      @logger.debug("Got response from Kubernetes API", response: response)
+
+      data = LogStash::Json.load(response.body)
+
+      {
+        'annotations' => sanitize_keys(data['metadata']['annotations']),
```

Comment:
You probably should make annotations optional since they often include vary wide of things hat people won't always want collected in their logs. And some components may have a lot of annotations, not all of which are useful and may just bloat the message (last applied config is a good example) Probably best to allow users to whitelist keys.